src/index.ts

```diff
@@ -0,0 +1,763 @@
+   *
+   * @param {string} nodeId
+   */
+  getPathForNodeId (nodeId) {
+    if (nodeId === this.rootNodeId) {
+      return this.rootNodeId
+    }
+    return this.nodePathList.find(path => path.endsWith('.' + nodeId))
+  }
+
+  /**
+   * register a new node into the LRUD tree
+   *
+   * @param {string} nodeId
+   * @param {object} node
+   * @param {string} [node.id] if null, `nodeId` is used
+   * @param {string} [node.parent] if null, value of `this.rootNodeId` is used
+   * @param {number} [node.index] if null, index is 1 more than the index of the last sibling. if no previous siblings, index is 1
+   * @param {number[]} [node.indexRange] defaults to null. acts as a colspan, value [0] is lower bound, value [1] is upper bound
+   * @param {function} [node.selectAction] if a node has a selectAction, it is focusabled
+   * @param {boolean} [node.isFocusable] if a node is explicitly set as isFocusable, it is focusable
+   * @param {boolean} [node.isWrapping] if true, when asking for the next child at the end or start of the node, the will "wrap around" and return the first/last (when asking for the last/first)
+   * @param {string} [node.orientation] can be "vertical" or "horizontal". is used in conjuction when handling direction of key press, to determine which child is "next"
+   * @param {boolean} [node.isIndexAlign] if a node is index aligned, its descendents should jump to nodes based on index instead of activeChild
+   * @param {function} [node.onLeave] if a node has an `onLeave` function, it will be run when a move event leaves this node
+   * @param {function} [node.onEnter] if a node has an `onEnter` function, it will be run when a move event enters this node
+   */
+  registerNode (nodeId, node : any = {}) {
+    if (!node.id) {
+      node.id = nodeId
+    }
+
+    // if this is the very first node, set it as root and return...
+    if (Object.keys(this.tree).length <= 0) {
+      this.rootNodeId = nodeId
+      this.tree[nodeId] = node
+      this.nodePathList.push(nodeId)
+      return this
+    }
+
+    // if this node DOESNT have a parent assume its parent is root
+    if (node.parent == null && nodeId !== this.rootNodeId) {
+      node.parent = this.rootNodeId
+    }
+
+    // if this node is the first child of its parent, we need to set its parent's `activeChild`
+    // to it so that the parent always has an `activeChild` value
+    // we can tell if its parent has any children by checking the nodePathList for
+    // entries containing '<parent>.children'
+    const parentsChildPaths = this.nodePathList.find(path => path.includes(node.parent + '.children'))
+    if (parentsChildPaths == null) {
+      const parentPath = this.getPathForNodeId(node.parent)
+      Set(this.tree, parentPath + '.activeChild', nodeId)
+    }
+
+    // if no `index` set, calculate it
+    if (!node.index) {
+      let parentNode = this.getNode(node.parent)
+      if (parentNode) {
+        let parentsChildren = this.getNode(node.parent).children
```

Comment:
Can `parentNode` and `parentsChildren` be `const` here?

Reply:
Indeed they can 👍 fixed.